### PR TITLE
Unit tests for the statemachine

### PR DESCRIPTION
**Pull request by oleorhagen**
All commits are related to unit-testing, and fixing up the state-machine (which is the ugliest part of the codebase atm, imo).

### CODE REVIEW

File: src/mender/statemachine/statemachine.py [hunk 1]
```diff
@@ -288,28 +289,36 @@ def run(self, context):
...
 
+class UnsupportedState(Exception):
+    pass
+
+
 class ArtifactReboot(State):
     def run(self, context):
         log.info("Running the ArtifactReboot state...")
-        return ArtifactCommit()
+        # return ArtifactCommit()
```
Comment (lluiscampos) [inline src/mender/statemachine/statemachine.py:299]:
```suggestion
        # return ArtifactReboot()
```
?

File: tests/unit/test_statemachine.py [hunk 2]
```diff
@@ -0,0 +1,264 @@
...
+            m.setattr(deployments, "request", lambda *args, **kwargs: {"foo": "bar"})
+            assert sync_update.run(ctx)
+
+    def test_download(self, monkeypatch, ctx):
+        download_state = statemachine.Download()
+        # Failed download
+        with monkeypatch.context() as m:
+            m.setattr(deployments, "download", lambda *args, **kwargs: None)
+            assert isinstance(download_state.run(ctx), statemachine.ArtifactFailure)
+        # Failed download
```
Comment (lluiscampos) [inline tests/unit/test_statemachine.py:149]:
```suggestion
        # Successful download
```
right?

File: tests/unit/test_statemachine.py [hunk 3]
```diff
@@ -0,0 +1,264 @@
...
+    def test_unathorized(self, ctx, monkeypatch):
+        unauthorized_machine = statemachine.UnauthorizedStateMachine()
+
+        class MockAuthorize:
+            def run(self, _):
+                return "JWTTOKENTEXT"
+
+        with monkeypatch.context() as m:
+            m.setattr(statemachine, "Authorize", MockAuthorize)
+            assert not unauthorized_machine.run(ctx)
```
Comment (lluiscampos) [inline tests/unit/test_statemachine.py:226]: This one I don't understand: it does not return anything so we should not assert on return value.

File: src/mender/statemachine/statemachine.py [hunk 4]
```diff
@@ -176,20 +179,17 @@ def run(self, context):
...
 
-# Should transitions always go through the external state-machine, to verify and
-# catch de-authorizations (?)
 #
 # Second layered machine (below Authorized)
 #
-# Idling - Or, just pushing inventory and identity data and looking for updates
 
 
 class SyncInventory(State):
```
Comment (lluiscampos) [inline src/mender/statemachine/statemachine.py:195]: This state does not retry on `client_inventory.request` failures (it ignores any return from there). I think we do retry on golang client - it is probably fine, I just wanted to note it.

**Review by lluiscampos — APPROVED** [review]
The tests are :top: class - they really helped following the full state machine logic.

Two minor comments below, one question for `SyncInventory` - but I _think_ it is correct.

File: src/mender/statemachine/statemachine.py (hunk 4 above)
Comment (oleorhagen) [reply src/mender/statemachine/statemachine.py:195]: Yes, correct observation. No retries, and no backoffs. Keeping it simple.

**Comment by merlin-northern** [comment]
since Lluis did the review I am not going to do a full-blown one. makes sense to me. and looks nice.

**Pull request merged**